### PR TITLE
docs: reputation portability design proposal

### PR DESCRIPTION
**Pull request by grunch**
## Summary

Design proposal for letting a user carry reputation earned on lnp2pBot, or on another Mostro instance, into a Mostro instance without giving any operator, or anyone reading relays, a way to link the two identities. Documentation only, no code changes.

`docs/REPUTATION_PORTABILITY.md` covers:

- **Threat model.** Source and destination operators, collusion or a leaked source database, relay observers. Two attacks drive the design: fingerprinting by exact values, and an issuer recognising what it signed.
- **Reputation bands.** A three-dimensional grid (completed trades, average rating, account age) as a protocol constant, seeded on the destination at the band floor, with a K-anonymity merge rule for sparse cells.
- **Blind signatures per band.** BDHKE as used by Cashu, one issuer key per grid cell, DLEQ proofs against a published keyset, epoch rotation. Mostro already depends on `cdk`, the bot can use `@cashu/crypto`, the 1.x app needs a small Dart implementation, the 2.x app gets it from `cdk` in its Rust core.
- **Merging on the destination.** Trades add up, rating is a weighted average, account age is a maximum and never a sum. Imported reputation lands in the ordinary fields so counterparties see one rating, one review count and one age.
- **`since` instead of `days`.** A day-truncated Unix timestamp replaces the derived day count at the three sites where it is exposed today (order `rating` tag, kind 38384 event, `UserInfo` in the `Peer` payload), with a deprecation window.
- **Mostro-to-Mostro.** Only native reputation is exportable, one token per issuer and destination identity, per-instance trust list of issuers.
- **Implementation plan.** Seven dependency-ordered phases across `protocol`, `mostro-core`, `mostro`, `mobile` (1.x), `app` (2.x) and `lnp2pbot/bot`, one PR per row with scope and acceptance criteria, aligned with mostro-core 0.14.6 (`Rating` has no date field, `User` derives `FromRow` over `SELECT *`, seed math belongs next to `User::update_rating`).
- **Open decisions.** Band cuts, K, eligibility, keyset event kind, re-issuance override, deprecation window length.

## Scope Declaration

- **Event domains affected:** none in this PR (documentation only). The proposal itself touches the order `rating` tag, the kind 38384 rating event and the `Peer` payload; those changes land in later PRs listed in section 9.
- **Cross-kind change:** no
- **External compatibility impact:** none from this PR. The proposed `since` field is additive during the deprecation window; `Rating::from_tags` ignores unknown keys and `UserInfo` would gain it as an optional field.

## Review focus

Concept review is what this PR needs: the threat model in section 3, the band grid in section 4, the merge rules in section 6 and the open decisions in section 10. The PR plan in section 9 is meant to be adjusted as decisions close.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01AataUN5L9RNdyyMZYm9jC9

---
_Generated by [Claude Code](https://claude.ai/code/session_01AataUN5L9RNdyyMZYm9jC9)_

## Summary by CodeRabbit

- **Documentation**
  - Added a design proposal for portable reputation between trading venues.
  - Documents privacy guarantees, threat models, reputation bands, and blind-signature workflows.
  - Defines encrypted export/import exchanges, identity validation, merge rules, timestamp handling, and cross-venue behavior.
  - Specifies canonical token encoding, validation rules, issuer keysets, replay protection, native account age, and reputation export timestamps.
  - Documents immutable keysets within each epoch and exact native reputation statistics.
  - Covers Sybil resistance, re-issuance constraints, phased implementation plans, and resolved design decisions.

### CODE REVIEW

File: docs/REPUTATION_PORTABILITY.md [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+
+## 10. Open decisions
+
+1. Band cuts for the three dimensions (section 4).
+2. `K` minimum cell population (default 50).
+3. Eligibility minimums (default 10 trades, disputes lost < 10%).
+4. Keyset event kind number and epoch length (default yearly).
+5. Admin override for re-issuance: yes or no.
+6. Length of the `days` deprecation window before PR 1.4 (default one minor release).
+7. ~~Whether the new actions bump `PROTOCOL_VER`~~ Closed: they do not. New actions ride on the unknown-variant tolerance already in the protocol; `PROTOCOL_VER` stays at 2.
```
[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md]: The premise is right and the conclusion is not, so `PROTOCOL_VER` stays at 2 with a different justification written into the document.

Confirmed against mostro-core 0.14.6: `#[serde(other)]` appears exactly once in the crate, on `CantDoReason::Unknown` in `src/error.rs`. `Action` and `Payload` have no unknown-variant fallback, so the original wording in this PR ("new actions ride on the unknown-variant tolerance already in the protocol") was wrong and has been removed.

What makes it safe is the message topology, not serde. These four actions are strictly request/response over targeted NIP-59 gift wraps:

- `export-reputation` and `import-reputation` are only ever sent by a client that implements them.
- `reputation-exported` and `reputation-imported` are only ever sent back to the client that made the request.
- Nothing here is broadcast; the order events that do go out to every client are tags, not `Message` values.

So no client running 0.14.6 is ever handed an `Action` or `Payload` variant it does not know. An old *daemon* receiving `import-reputation` from a new client fails to deserialise and errors, which is the correct degradation for a feature it does not implement.

Section 9 now carries this reasoning as a "Protocol compatibility" note, and PR 2.3 in the plan carries a test pinning it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by grunch** [comment]
Thanks — the review caught a fatal flaw in the original design. All findings are addressed in the latest commit; summary of what changed and why.

## The scheme was unimplementable, and is replaced

Both reviewers independently flagged that the destination cannot verify `C == k_cell·hash_to_curve(secret)` given only `K_cell`. That is correct and fundamental: in BDHKE the *mint* verifies its own tokens with the private key, and deciding `C = k·Y` from `K = k·G` alone is DDH, assumed hard on secp256k1. Our issuer and verifier are different parties, so BDHKE cannot work here at all.

Replaced with **blind Schnorr on secp256k1, clause variant** (Fuchsbauer–Kiltz–Loss, for ROS resistance under concurrent sessions). It is blind at issuance and publicly verifiable at redemption: the destination checks `s·G == R + H(R‖m)·P_cell` against the published keyset, using only public data. Consequences:

- Issuance becomes a two-round-trip stateful session (open two nonce clauses, answer one at random), so both issuer and client persist session state; an abandoned session expires without consuming the once-only flag.
- `cdk` is no longer relevant. Curve arithmetic comes from `k256` in mostro and in the 2.x app's Rust core, `@noble/curves` in the bot, and a small Dart implementation in the 1.x app.
- The client now verifies the signature itself before storing the token, which also replaces the DLEQ proof as the defence against an issuer signing with an off-keyset key.

## Trades vs reviews

Correct and serious. `total_reviews` only increments on an actual submitted rating, so seeding it from a completed-trade count would publish reviews the user never earned and give one rating the weight of hundreds. The first band dimension now counts **ratings received**; completed trades decide eligibility only. The worked example was rebuilt accordingly (347 trades, 214 ratings received → a 200-review seed).

## Cell-selection tagging

Correct, and DLEQ never covered it. Added an explicit client-side check: the user knows their own review count, rating and account age, so the client recomputes the cell and aborts if the issuer named a different one. The unauditable `K` population remains a residual risk and is documented as such rather than claimed solved.

## Keyset `d` tag

Good catch. With a fixed `d`, NIP-01 replacement semantics would delete the previous epoch's keyset from relays while the destination still accepts tokens from it. The epoch is now part of the `d` value.

## Band intervals

All bands are now half-open `[low, high)` with explicit floors, a month defined as exactly 30 × 86400 s, and ages computed from the day-truncated timestamp. The top rating band is closed at 5.0 since that is the maximum a rating can take. Boundary values are now in the test-vector requirements for PR 0.2/2.1.

## Unlinkability wording

Fair — goal 3 claimed an absolute guarantee while section 3 admitted a timing residual. Goal 3 now states cryptographic unlinkability *on the protocol data*, explicitly probabilistic against traffic analysis, and section 3 bounds the residual: the anonymity set is the number of migrations from the same issuer inside the random redemption-delay window, small on day one, and nothing helps the very first migrant.

## Multi-destination vs destination binding

Correct contradiction. Resolved by choosing Sybil resistance: **one token, one destination identity, redeemable on any instance that trusts the issuer**. The "fresh identity per instance" option is dropped, with the reason recorded in section 7 — supporting it needs one token per identity, and nothing could then stop two of them being redeemed under two identities on the same instance, which is exactly the property the design is supposed to provide. The corresponding PRs (5.7, 5b.6) are removed from the plan.

## `PROTOCOL_VER`

Answered inline on that thread: the premise is right, the conclusion is not. Staying at 2, justified by request/response topology rather than by a serde tolerance that only `CantDoReason` has.

## MD040

Fixed; `markdownlint-cli2@0.23.2` now reports 0 issues across the repo.

Section 10 records which decisions were closed by this round and which remain open (band cuts, `K`, eligibility minimums, keyset kind number, re-issuance override, deprecation window, delay window width).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by grunch** [comment]
Second round addressed — all seven findings were valid, two of them substantive design bugs rather than spec gaps.

**Canonical encoding** (two threads). The spec used symbolic `‖` with no byte definition, which would have let the Rust, JS and Dart implementations derive different challenges for the same token. New section 5.1 pins it: 33-byte compressed SEC1 points, 32-byte big-endian scalars rejected if zero, BIP-340-style tagged hashes, `c = int(H_"mostro/reputation/challenge/v1"(R ‖ m)) mod n`, a fixed-length 70-byte `m` (`"repv1:"` + 32-byte x-only destination pubkey + 32-byte nonce) and a tagged token id that replaces the vague `hash(m)`. Every field is fixed length, so concatenation needs no extra framing.

Worth flagging one thing the encoding section now states explicitly: points are compressed rather than x-only, and the challenge is a plain tagged hash rather than BIP-340's. BIP-340 normalises `R` to even Y, and the client's blinded `R_i = R'_i + α_i·G + β_i·P_cell` has unpredictable parity that the signer cannot compensate for, so reusing BIP-340 verbatim would break the blinding. PR 0.5 now requires encoding vectors alongside the transcript.

**K-anonymity merging vs client validation.** Genuine contradiction introduced by the previous round: the issuer folds sparse cells downward, the client recomputes the raw cell and aborts on any mismatch, so every user in a merged cell would have been unable to complete issuance. Fixed by making the merge deterministic (step down one band, dimensions tried in the fixed order ratings → age → rating, repeat until the absorber holds `K`) and publishing the raw→effective map as `merges` in the keyset event. The client applies the same map transitively before comparing. Publishing the map is also strictly better than the original design, since the client can now check the issuer actually followed the rule instead of trusting it.

**Native account age.** Real bug, and the best catch of the round. `created_at` is backdated on import while the export path read it for the age band, so after A→B, instance B would have exported A's age to C as its own native age — seeds travelling twice on exactly the dimension where "only native reputation is exportable" was supposed to hold. Added `native_created_at`, set once at row creation and never moved; exports read it, the displayed `created_at` keeps the merged value. Reviews and rating were already protected by `seeded_reviews` / `seeded_rating_sum`; age needed its own field because its merge is a minimum rather than a subtraction. PR 2.5 now carries an A→B→C property test.

**`reputation_exported_at` precision.** Section 3 already promised "no issuance timestamp finer than the per-user flag", but the field was typed `Option<i64>` with no precision stated. Now day-truncated with the same helper as `since`, and the necessarily fine-grained session state is deleted on completion or expiry.

**Issuer identity.** `NOSTR_SK` in section 5.1 vs `REPUTATION_ISSUER_SK` in PR 6.1 could have made the keyset author, the trusted-list entry and the token `issuer` field three different authorities. There is now one issuer identity per issuer; for the bot it is `REPUTATION_ISSUER_SK`, deliberately separate from `NOSTR_SK` so reputation issuance can be rotated or revoked without touching the bot's other Nostr activity.

**Action naming.** `reputation-exported` is the single wire discriminator, used in the schema, the core enum serialization, the compatibility note and the vectors. `export-reputation-response` is gone.

`markdownlint-cli2@0.23.2` reports 0 issues. Section 10 records the decisions closed by this round.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/REPUTATION_PORTABILITY.md [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+
+| Element | Encoding |
+|---|---|
+| Curve points | 33-byte compressed SEC1 (`0x02`/`0x03` prefix) |
+| Scalars | 32-byte big-endian, reduced mod `n`, rejected if zero |
+| Tagged hash | `H_tag(x) = SHA256(SHA256(tag) ‖ SHA256(tag) ‖ x)`, as in BIP-340 |
+| Challenge | `c = int(H_"mostro/reputation/challenge/v1"(R ‖ m)) mod n`, `R` compressed |
+| `m` | `"repv1:"` (6 ASCII bytes) ‖ 32-byte x-only destination identity pubkey ‖ 32-byte random nonce — 70 bytes exactly |
+| Token id | `H_"mostro/reputation/token/v1"(m)`, the primary key in `redeemed_reputation_tokens` |
+| Cell id | UTF-8 `reviews:<label>\|rating:<label>\|age:<label>`, no spaces, labels from the section 4 table |
```
[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md:157]: Added in efee4b4: section 5.1 now carries the token wire table (`issuer` 32-byte hex, `epoch`, `cell`, `m` 70-byte hex, `r_point` 33-byte hex, `s` 32-byte hex), the rejection rules (missing/duplicate/wrong length/not lowercase hex/off-curve/zero or ≥ n), and the note that nothing is hashed over the JSON. PR 0.5 now includes one complete serialised token with its decoded bytes and token id. The normative schema itself stays in the protocol repo (PR 0.4), as before.

File: docs/REPUTATION_PORTABILITY.md [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+all three sites, then drops `days`. `Rating::from_tags` ignores unknown keys
+and `UserInfo` gains the field as `Option` with a serde default, so old and
+new peers interoperate during the window.
+
+## 7. Mostro-to-Mostro specifics
+
+- **Only native reputation is exportable.** When a Mostro acts as issuer it
+  computes the cell from `total_reviews - seeded_reviews`, the native rating
+  derived from `seeded_rating_sum`, and `native_created_at` rather than the
+  backdated `created_at`. Seeds never travel twice, which
```
[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md]: Addressed in efee4b4 with an accumulator rather than a reversal: `users.native_rating_sum` is incremented by `update_rating` on native reviews only and never touched by `apply_reputation_seed`, so `native_rating = native_rating_sum / (total_reviews - seeded_reviews)` (0 when there are no native reviews). Section 7 spells out why reversing `total_rating` is not exact (f64 running mean plus the first-vote 1/2 rule). PR 2.5 lists the tests: one native review, several, seed then native, several seeds, native before and after a seed. PR 3.2 backfills the column for legacy rows.

File: docs/REPUTATION_PORTABILITY.md [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+| 2.4 | mostro-core | `src/user.rs`: `User` gains `seeded_reviews: i64`, `seeded_rating_sum: f64`, `native_created_at: i64`, `reputation_exported_at: Option<i64>` (day-truncated), each with `#[sqlx(default)]` and `#[serde(default)]` so a daemon on an un-migrated database still deserialises `SELECT *`. `User::new` initialises `native_created_at` to `created_at`. | First use of `sqlx(default)` in core; test with a row that lacks the columns; existing rows backfill `native_created_at` from `created_at`. |
+| 2.5 | mostro-core | `src/user.rs`: `User::apply_reputation_seed(&mut self, cell: &Cell, now: i64)` implementing section 6 next to `update_rating`, plus `User::native_stats(&self) -> (reviews, rating, since)` that subtracts the seeded part and reads `native_created_at`, never `created_at`. Property tests: never decreases `total_reviews`, never moves `created_at` forward, never touches `native_created_at`, and an A→B→C chain exports the same native stats B had before importing from A. | No I/O; released as **0.15.0** together with 2.1-2.4. |
```
[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md]: Fixed in efee4b4: `native_created_at` is now `Option<i64>`, `User::new` sets `Some(created_at)`, and readers go through `User::native_created_at()` which falls back to `created_at`. PR 2.4's test now asserts the fallback equals `created_at`, never 0.

File: docs/REPUTATION_PORTABILITY.md [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+| 3.3 | mostro | Crypto module `reputation::verify` over `k256` (new dependency), using `reputation::encoding` from core so the byte contract is shared: challenge hash and the `s·G == R + H(R‖m)·P_cell` check. Tested against the 0.5 vectors, including every invalid case. | No daemon wiring yet. |
+| 3.4 | mostro | Keyset fetcher: fetch and cache the issuer keyset event per trusted issuer, refresh on epoch change, reject unknown epochs. Tested with the `local-relay` feature. | Cache survives a relay outage. |
+| 3.5 | mostro | Handler `src/app/import_reputation.rs`: routing in `app.rs`, checks 1-5 of section 5.4 in one transaction, calls `User::apply_reputation_seed` from core, persists through a new `update_user_reputation_seed` in `db.rs`, replies `reputation-imported` or `cant-do`. Integration test end to end with a fixture issuer. | A second redemption of the same token is rejected. |
```
[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md]: Fixed in efee4b4.

File: docs/REPUTATION_PORTABILITY.md [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+| 3.6 | mostro | Publish the updated kind 38384 rating event after a successful import, reusing `update_user_rating_event`. | Event visible on the local relay. |
+
+### Phase 4: mostrod as issuer (export)
+
+| PR | Repo | Scope | Done when |
+|---|---|---|---|
+| 4.1 | mostro | Settings `[reputation_export]` (`enabled`); per-cell key derivation from the daemon key and the protocol-constant epoch label (HKDF, deterministic, never stored). Unit tests: same inputs, same keys; a different epoch label yields different keys. | No publication yet. |
+| 4.2 | mostro | Publish the keyset event at startup and on epoch rollover: cell populations from the database, the deterministic merge from section 4, and both `cells` and the raw→effective `merges` map. | Event validates against 2.2; a client applying `merges` reproduces the issuer's assignment for every raw cell. |
```
[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md]: Addressed in efee4b4. Section 5.2 now states that a keyset is immutable within its epoch: the merge is computed once when the epoch opens, the raw→effective map is persisted (`reputation_keysets` table in 4.2, a Mongo collection in 6.2), and every later startup republishes byte-identical `cells` and `merges`. Population changes only shape the next epoch. Clients treat a changed `cells`/`merges` under the same `d` as an issuer fault. Went with this instead of a keyset revision in the token to keep the token and the accepted-epochs rule unchanged.

[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md]: On the HKDF point in this same thread, also fixed in efee4b4: PR 4.1 now specifies `HKDF(ikm = daemon secret key, salt = "mostro/reputation/keyset/v1", info = epoch label ‖ 0x00 ‖ cell id)` with the canonical section 5.1 cell id, and the tests include "different cell id → different key" and "no duplicate P_cell across the grid". 6.1 references the same scheme.

File: docs/REPUTATION_PORTABILITY.md [hunk 7]
```diff
@@ -0,0 +1,636 @@
...
+| `m` | `"repv1:"` (6 ASCII bytes) ‖ 32-byte x-only destination identity pubkey ‖ 32-byte random nonce — 70 bytes exactly |
+| Token id | `H_"mostro/reputation/token/v1"(m)`, the primary key in `redeemed_reputation_tokens` |
+| Cell id | UTF-8 `reviews:<label>\|rating:<label>\|age:<label>`, no spaces, labels from the section 4 table |
+
+The token travels as the `ReputationToken` payload of `import-reputation`
+(PR 0.4 owns the normative schema; this is the byte contract it must state).
+Binary fields are lowercase hex, and a parser rejects the token before any
+curve arithmetic when a field is missing, duplicated, of the wrong length,
+not lowercase hex, or when a point does not decode onto the curve or a scalar
+is zero or not below `n`:
```
[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md]: Fixed in 60def37: the 5.1 table now defines wire scalars as canonical integers in `1..n-1`, rejected (never reduced) when zero or `≥ n`; reduction mod n is confined to arithmetic. The 0.5 invalid-token table gains `s = 0` and `s = n + v` (must be rejected, not accepted as `v`).

File: docs/REPUTATION_PORTABILITY.md [hunk 8]
```diff
@@ -0,0 +1,636 @@
...
+
+| Field | Type | Bytes | Meaning |
+|---|---|---|---|
+| `issuer` | hex | 32 | x-only pubkey of the issuer's identity key, the one that signed the keyset |
+| `epoch` | string | – | epoch label, must match the keyset's `epoch` tag |
+| `cell` | string | – | effective cell id, must be a key of the keyset's `cells` |
```
[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md:170]: Fixed in 60def37: section 5.2 now lists the acceptance rules a consumer applies before reading any `P_cell` — valid Nostr signature, `pubkey` equal to the trusted issuer key, reserved kind, `d` equal to `reputation-keyset:<epoch>` with a matching `epoch` tag, and well-formed `cells`/`merges`. Failing events are discarded and never cached. PR 3.4 implements and tests it (wrong signer, `d`/`epoch` mismatch).

File: docs/REPUTATION_PORTABILITY.md [hunk 9]
```diff
@@ -0,0 +1,636 @@
...
+
+### Phase 3: mostrod as destination (import)
+
+| PR | Repo | Scope | Done when |
+|---|---|---|---|
+| 3.1 | mostro | Settings section `[reputation_import]` (`enabled`, `issuers`, `accepted_epochs`) with parsing, defaults and validation. No behaviour. | Bad config is rejected at startup with a clear error. |
+| 3.2 | mostro | Bump core to 0.15.0. Migration `users` gains `seeded_reviews`, `seeded_rating_sum`, `native_rating_sum` (backfilled as `total_rating * total_reviews`, the best available estimate for legacy rows), `native_created_at` (backfilled from `created_at`) and `reputation_exported_at` (matching PR 2.4 exactly, `SELECT *` + `FromRow` requires it); new table `redeemed_reputation_tokens(token_id PK, issuer, identity_pubkey, cell, redeemed_at)` with a unique index on `(issuer, identity_pubkey)`. `db.rs` accessors with tests. | Migration applies on an existing database; the `users` insert in `db.rs` binds the new columns. |
```
[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md]: Kept the backfill, reworded the rationale in 60def37. Mostro stores no per-review history (`update_rating` in core 0.14.6 keeps only the running mean), so there is nothing to replay. With `seeded_reviews = 0` the section 7 division returns `total_rating * total_reviews / total_reviews`, i.e. the displayed rating exactly, so a legacy user exports the reputation they show today — which is the intended cell. Marking legacy rows as unknown would block export for every existing user, defeating the feature's purpose.

File: docs/REPUTATION_PORTABILITY.md [hunk 10]
```diff
@@ -0,0 +1,636 @@
...
+| 3.5 | mostro | Handler `src/app/import_reputation.rs`: routing in `app.rs`, checks 1-5 of section 5.4 in one transaction, calls `User::apply_reputation_seed` from core, persists through a new `update_user_reputation_seed` in `db.rs`, replies `reputation-imported` or `cant-do`. Integration test end-to-end with a fixture issuer. | A second redemption of the same token is rejected. |
+| 3.6 | mostro | Publish the updated kind 38384 rating event after a successful import, reusing `update_user_rating_event`. | Event visible on the local relay. |
+
+### Phase 4: mostrod as issuer (export)
+
+| PR | Repo | Scope | Done when |
+|---|---|---|---|
+| 4.1 | mostro | Settings `[reputation_export]` (`enabled`); per-cell key derivation `x_cell = HKDF(ikm = daemon secret key, salt = "mostro/reputation/keyset/v1", info = epoch label ‖ 0x00 ‖ cell id)` (deterministic, never stored), the cell id being the canonical section 5.1 string so each cell gets its own scalar. Unit tests: same inputs, same keys; a different epoch label yields different keys; a different cell id yields different keys; the derived `P_cell` set has no duplicates across the grid. | No publication yet. |
```
[the remark this answers is not in the document]
Comment (grunch) [reply docs/REPUTATION_PORTABILITY.md]: Addressed in 60def37: PR 4.1 now pins HKDF-SHA256 (RFC 5869), `ikm` = 32 big-endian secret-key bytes, the ASCII salt, the `info` layout (epoch ‖ 0x00 ‖ cell id ‖ counter byte), `L = 32`, and the counter-retry rule for an output that is 0 or ≥ n. PR 6.1 references that exact contract and PR 0.5 gains a derivation vector so both implementations test the same scheme. Note the two issuers still derive from different secrets, so identical *keys* across them is not a goal — identical *scheme* is.
